### DISCUSSION TITLE
Testing Router with State

### DISCUSSION THREAD

**Discussion by McPatate** [tokio-rs/axum#1658] [Q&A] [answered]
Hello, I've been trying to no avail to set up tests for a Router which has a state.

From reading the source code in `routing.rs`, it seems that the trait bounds for `tower::Service[Ext]` that enable calls to `.oneshot()` and `.call()` are only implemented for a `Router<()>` and not for `Router<S>`.

https://github.com/tokio-rs/axum/blob/978ae6335862b264b4368e01a52177d98c42b2d9/axum/src/routing/mod.rs#L553

Code example of what I'm attempting to do :

```rs
#[derive(Clone)]
struct Container  {
    /// some fields
}

async fn handler(container: State<Container>) {
     /// ...
}

fn router() -> Router<Container> {
    Router::new().route("/", get(handler))
}

#[cfg(test)]
mod tests {
    use super::*;
    use axum::{
        body::Body,
        http::{self, Request, StatusCode},
    };
    use tower::ServiceExt;

    #[tokio::test]
    async fn test_route() {
        let app = router();

        let response = app
            .oneshot(
                Request::builder()
                    .method(http::Method::GET)
                    .uri("/")
                    .body(Body::empty())
                    .unwrap()
            )
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);
    }
}
``` 

The content of my State is `dyn trait`s, the goal here being to mock my database repositories.

Is what I'm doing sound ? Or should I use Extensions for this use case ?

**Comment by davidpdrsn** [comment, marked as answer]
You need to call https://docs.rs/axum/latest/axum/routing/struct.Router.html#method.with_state

**Reply by McPatate** [reply]
Do you mean in my test case ? I'm still getting the error

**Reply by McPatate** [reply]
For context, the router is merged in another "main" router that has the `.with_state()` call

**Reply by davidpdrsn** [reply]
> Do you mean in my test case ? I'm still getting the error

Yes. You need to provide the state, otherwise you cannot call the router. How would it handle requests if it doesn't have the state?

> For context, the router is merged in another "main" router that has the .with_state() call

Make sure you're specifying the right router type. See https://docs.rs/axum/latest/axum/routing/struct.Router.html#method.with_state for more details.

**Reply by McPatate** [reply]
The merge works without issue, where I'm having trouble is in my test case. Even when adding the state it still errors at compile :
```rs
    #[tokio::test]
    async fn test_route() {
        let app = router();
        let container = Container {};
        app.with_state(container);

        let response = app
            .oneshot(
                Request::builder()
                    .method(http::Method::GET)
                    .uri("/")
                    .body(Body::empty())
                    .unwrap()
            )
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);
    }
```
```
error[E0599]: the method `oneshot` exists for struct `Router<container::Container>`, but its trait bounds were not satisfied
    |
125 |             .oneshot(
    |              ^^^^^^^
    |
   ::: .cargo/registry/src/github.com-1ecc6299db9ec823/axum-0.6.1/src/routing/mod.rs:64:1
    |
64  | pub struct Router<S = (), B = Body> {
    | -----------------------------------
    | |
    | doesn't satisfy `Router<container::Container>: Service<_>`
    | doesn't satisfy `Router<container::Container>: tower::ServiceExt<_>`
    |
    = note: the following trait bounds were not satisfied:
            `Router<container::Container>: Service<_>`
            which is required by `Router<container::Container>: tower::ServiceExt<_>`
            `&Router<container::Container>: Service<_>`
            which is required by `&Router<container::Container>: tower::ServiceExt<_>`
            `&mut Router<container::Container>: Service<_>`
            which is required by `&mut Router<container::Container>: tower::ServiceExt<_>
```

**Reply by McPatate** [reply]
Ah, my bad, this seems to fix things :
```rs
    let app = app.with_state(container);
```
Or calling `.with_state()` directly on the `router()` call. Anyhow I think I understand what was wrong.

Thanks for helping me out on this !
